cmd/operator-sdk/generate/bundle/cmd.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2020 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bundle
+
+import (
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	kbutil "github.com/operator-framework/operator-sdk/internal/util/kubebuilder"
+)
+
+//nolint:maligned
+type bundleCmd struct {
+	// Options to turn on different parts of bundling.
+	kustomize bool
+	manifests bool
+	metadata  bool
+
+	// Common options.
+	operatorName string
+	version      string
+	inputDir     string
+	outputDir    string
+	manifestRoot string
+	apisDir      string
+	crdsDir      string
+	stdout       bool
+	quiet        bool
+
+	// Metadata options.
+	channels       string
+	defaultChannel string
+	overwrite      bool
+}
+
+//nolint:lll
+func NewCmd() *cobra.Command {
+	c := &bundleCmd{}
+	cmd := &cobra.Command{
+		Use:   "bundle",
+		Short: "Generates bundle data for the operator",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 0 {
+				return fmt.Errorf("command %s doesn't accept any arguments", cmd.CommandPath())
+			}
+
+			// Turn them all on by default here to simplify individual flag usage.
+			fs := cmd.Flags()
+			if !fs.Changed("kustomize") && !fs.Changed("metadata") && !fs.Changed("manifests") {
+				c.kustomize = true
+				c.manifests = true
+				c.metadata = true
```

Comment:
If all is by default true so, why not set the default value equals true when the flag is defined [here](https://github.com/operator-framework/operator-sdk/pull/2860/files#diff-0aff0dcd8bfc140fd01b40d46a1f7bc1R99)?

Reply:
If no `kustomize`, `metadata`, or `manifests` flag is set, this command does nothing. The command is intended to generate a bundle, a process involving all 3 of these steps. Therefore, if the user sets no flag, I take that to mean "do the thing this command is supposed to". Turning off functionality with a boolean flag can be annoying if there are too many booleans:

If flag defaults are set to true:

```sh
# To only generate metadata
$ operator-sdk generate bundle --kustomize=false --manifests=false
# To generate kustomize and metadata
$ operator-sdk generate bundle --manifests=false
# To generate kustomize, manifests, metadata
$ operator-sdk generate bundle
```

If flag defaults are set to false:

```sh
# To only generate metadata
$ operator-sdk generate bundle --metadata
# To generate kustomize and metadata
$ operator-sdk generate bundle --metadata --kustomize
# To generate kustomize, manifests, metadata
$ operator-sdk generate bundle --metadata --kustomize --manifests
```

If options are set to true in `RunE` (current):

```sh
# To only generate metadata
$ operator-sdk generate bundle --metadata
# To generate kustomize and metadata
$ operator-sdk generate bundle --kustomize --metadata
# To generate kustomize, manifests, metadata
$ operator-sdk generate bundle
```

I prefer the latter, wdyt?

Reply:
I prefer the latter, too/

Reply:
Actually looking your explanation we do not really care about the bool values. Am I right? I mean, if it has the flag --metadata than should gen metadata, if has the --kustomize --metadata should gen both and we follow in this way.

So, could not it be simplified by checking if the flag was or not used instead of to set true or false in the middle of the scenarios as done above? if you see any reasons for not, wdyt about at least add a comment on top of this if and in any other place that you set true/false to manipulate it in order to described why/purpose.

Reply:
If I'm understanding you correctly, you can still set `--kustomize=false`, so only checking usage isn't enough. Can you provide an example of what you mean?

Reply:
Just to clarifies. Note that, I am as a user I'd expected:

- To have a "bundle" with ALL requirements: > `$ operator-sdk generate bundle`
- To disable some option to gen this "bundle": `$ operator-sdk generate bundle --skip-option` OR `$ operator-sdk generate bundle --option=false`.  

Currently, we have, for example:  

> cmd.Flags().BoolVar(&c.metadata, "metadata", false, "Generate bundle metadata and Dockerfile")

According to the above explanation means: 

```
> # To only generate the metadata
> $ operator-sdk generate bundle --metadata
```

The above command would lead to me think that it will do **a full bundle +/with a metadata option**. However, the command will **ONLY** exec this specific piece instead of it. 

So, it is all fine if the current behaviour make sense for you folks . And then, in this situation, my suggestion would just improve the code comment and flags description to make its behaviour clear enough. E.g : 

```go
// Turn them all on by default here to simplify individual flag usage.
// # To only generate metadata
// $ operator-sdk generate bundle --metadata
// # To generate kustomize and metadata
// $ operator-sdk generate bundle --kustomize --metadata
// # To generate kustomize, manifests, metadata
// $ operator-sdk generate bundle
fs := cmd.Flags()
if !fs.Changed("kustomize") && !fs.Changed("metadata") && !fs.Changed("manifests") {
```
AND

```go
 // PS.: we need to clarifies it. Otherwise, is not clear why not just run `make kustomize instead of it`
cmd.Flags().BoolVar(&c.kustomize, "kustomize", false, "If set, it will generate/update ONLY the kustomize config bases with the specific bundle manifests")
cmd.Flags().BoolVar(&c.manifests, "manifests", false, "If set, it will generate/update ONLY the bundle manifests directory")
cmd.Flags().BoolVar(&c.metadata, "metadata", false, "If set, it will generate/update the bundle metadata directory and its Dockerfile")
```